File: 74-search-a-2d-matrix/74-search-a-2d-matrix.cpp
class Solution {
public:
    bool searchMatrix(vector<vector<int>>& matrix, int target) {
        //Approach_02: Binary Search: Time: O(log(n*m)) and Space:O(1)
        int n = matrix.size();
        int m = matrix[0].size();
        int low = 0, high = n*m-1;
        
        if(!matrix.size())
            return false;
        
        while(low <= high)
        {
            int middle = (low+(high-low)/2);
            //checking all the three condition for middle element:
            int mid_ele = matrix[middle/m][middle%m];
            if(mid_ele > target)
                high = middle-1;
            else if(mid_ele < target)
                low = middle+1;
            else
                return true;
        }
        return false;
    }
};

File: 74-search-a-2d-matrix/74-search-a-2d-matrix.java
class Solution {
    public boolean searchMatrix(int[][] matrix, int target) {
        int n = matrix.length;
        int m = matrix[0].length;
        
        if(matrix.length == 0)
            return false;
        int i = 0,j = m*n-1;
        
        while(i<=j)
        {
            int mid = (i+j)/2;
            if(matrix[mid/m][mid%m] == target)
                return true;
            else if(matrix[mid/m][mid%m] > target)
                j = mid-1;
            else
                i = mid+1;
        }
        return false;
    }
}